File: app/src/databases/neo4j.js
const neo4j = require('neo4j-driver');
const logger = require('../utils/logger');

let driver;

function getDriver() {
  if (!driver) {
    const uri = process.env.NEO4J_URI || 'bolt://localhost:7687';
    const user = process.env.NEO4J_USER || 'neo4j';
    const password = process.env.NEO4J_PASSWORD || 'logistics_pass';

    driver = neo4j.driver(uri, neo4j.auth.basic(user, password), {
      connectionTimeout: 10000,
      maxConnectionPoolSize: 10,
    });
  }
  return driver;
}

async function connect() {
  const d = getDriver();
  await d.verifyConnectivity();
  logger.info('Neo4j connection verified');
}

async function runQuery(cypher, params = {}) {
  const session = getDriver().session();
  try {
    const result = await session.run(cypher, params);
    return result.records;
  } finally {
    await session.close();
  }
}

async function close() {
  if (driver) {
    await driver.close();
    driver = null;
    logger.info('Neo4j driver closed');
  }
}

module.exports = { connect, runQuery, close };

File: app/src/databases/neo4j.py
from neo4j import GraphDatabase
from src.config.settings import settings
from src.utils.logger import get_logger

logger = get_logger(__name__)

_driver = None


def get_driver():
    global _driver
    if _driver is None:
        _driver = GraphDatabase.driver(
            settings.neo4j_uri,
            auth=(settings.neo4j_user, settings.neo4j_password),
        )
    return _driver


def connect():
    get_driver().verify_connectivity()
    logger.info("Neo4j connection verified")


def run_query(cypher: str, params: dict = None):
    with get_driver().session() as session:
        result = session.run(cypher, params or {})
        return [record for record in result]


def close():
    global _driver
    if _driver:
        _driver.close()
        _driver = None
        logger.info("Neo4j driver closed")
